SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialRPCService.cpp

```diff
@@ -0,0 +1,430 @@
+		}
+	}
+
+	return Components;
+}
+
+void SpatialRPCService::ExtractRPCsForEntity(Worker_EntityId EntityId, Worker_ComponentId ComponentId)
+{
+	switch (ComponentId)
+	{
+	case SpatialConstants::CLIENT_ENDPOINT_COMPONENT_ID:
+		if (View->HasAuthority(EntityId, SpatialConstants::SERVER_ENDPOINT_COMPONENT_ID))
+		{
+			ExtractRPCsForType(EntityId, ERPCType::ServerReliable);
+			ExtractRPCsForType(EntityId, ERPCType::ServerUnreliable);
+		}
+		break;
+	case SpatialConstants::SERVER_ENDPOINT_COMPONENT_ID:
+		if (View->HasAuthority(EntityId, SpatialConstants::CLIENT_ENDPOINT_COMPONENT_ID))
+		{
+			ExtractRPCsForType(EntityId, ERPCType::ClientReliable);
+			ExtractRPCsForType(EntityId, ERPCType::ClientUnreliable);
+		}
+		break;
+	case SpatialConstants::MULTICAST_RPCS_COMPONENT_ID:
+		ExtractRPCsForType(EntityId, ERPCType::NetMulticast);
+		break;
+	}
+}
+
+void SpatialRPCService::OnCheckoutEntity(Worker_EntityId EntityId)
+{
+	const MulticastRPCs* Component = View->GetComponentData<MulticastRPCs>(EntityId);
+	// When checking out entity, ignore multicast RPCs that are already on the component.
+	LastSeenMulticastRPCIds.Add(EntityId, Component->MulticastRPCBuffer.LastSentRPCId);
+}
+
+void SpatialRPCService::OnRemoveEntity(Worker_EntityId EntityId)
+{
+	LastSeenMulticastRPCIds.Remove(EntityId);
+}
+
+void SpatialRPCService::OnEndpointAuthorityGained(Worker_EntityId EntityId, Worker_ComponentId ComponentId)
+{
+	switch (ComponentId)
+	{
+	case SpatialConstants::CLIENT_ENDPOINT_COMPONENT_ID:
+	{
+		const ClientEndpoint* Endpoint = View->GetComponentData<ClientEndpoint>(EntityId);
+		LastAckedRPCIds.Add(EntityRPCType(EntityId, ERPCType::ClientReliable), Endpoint->ReliableRPCAck);
+		LastAckedRPCIds.Add(EntityRPCType(EntityId, ERPCType::ClientUnreliable), Endpoint->UnreliableRPCAck);
+		LastSentRPCIds.Add(EntityRPCType(EntityId, ERPCType::ServerReliable), Endpoint->ReliableRPCBuffer.LastSentRPCId);
+		LastSentRPCIds.Add(EntityRPCType(EntityId, ERPCType::ServerUnreliable), Endpoint->UnreliableRPCBuffer.LastSentRPCId);
+		break;
+	}
+	case SpatialConstants::SERVER_ENDPOINT_COMPONENT_ID:
+	{
+		const ServerEndpoint* Endpoint = View->GetComponentData<ServerEndpoint>(EntityId);
+		LastAckedRPCIds.Add(EntityRPCType(EntityId, ERPCType::ServerReliable), Endpoint->ReliableRPCAck);
+		LastAckedRPCIds.Add(EntityRPCType(EntityId, ERPCType::ServerUnreliable), Endpoint->UnreliableRPCAck);
+		LastSentRPCIds.Add(EntityRPCType(EntityId, ERPCType::ClientReliable), Endpoint->ReliableRPCBuffer.LastSentRPCId);
+		LastSentRPCIds.Add(EntityRPCType(EntityId, ERPCType::ClientUnreliable), Endpoint->UnreliableRPCBuffer.LastSentRPCId);
+		break;
+	}
+	case SpatialConstants::MULTICAST_RPCS_COMPONENT_ID:
+	{
+		const MulticastRPCs* Component = View->GetComponentData<MulticastRPCs>(EntityId);
+		LastSentRPCIds.Add(EntityRPCType(EntityId, ERPCType::NetMulticast), Component->MulticastRPCBuffer.LastSentRPCId);
+
+		// Duplicate the initial multicast RPCs?
```

Comment:
valentyn: Add a field with the number of RPCs initially put onto this component and copy that number to last sent ID on gaining authority.